akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/AccountExample2.java

```diff
@@ -0,0 +1,208 @@
+      return (balance - amount) >= 0.0;
+    }
+
+    private boolean isCloseable() {
+      return balance == 0.0;
+    }
+  }
+
+  public enum ClosedAccount implements Account {
+    INSTANCE
+  }
+
+  public static Behavior<AccountCommand> behavior(String accountNumber) {
+    return Behaviors.setup(context -> new AccountExample2(context, accountNumber));
+  }
+
+  public AccountExample2(ActorContext<AccountCommand> context, String accountNumber) {
+    super(new PersistenceId(accountNumber));
+  }
+
+  @Override
+  public Account emptyState() {
+    return EmptyAccount.INSTANCE;
+  }
+
+  private CommandHandlerBuilder<AccountCommand, AccountEvent, EmptyAccount, Account> initialHandler() {
+    return commandHandlerBuilder(EmptyAccount.class)
+      .matchCommand(CreateAccount.class, this::create);
+  }
+
+  private CommandHandlerBuilder<AccountCommand, AccountEvent, OpenedAccount, Account> openedAccountHandler() {
+    return commandHandlerBuilder(OpenedAccount.class)
+        .matchCommand(Deposit.class, this::deposit)
+        .matchCommand(Withdraw.class, this::withdraw)
+        .matchCommand(CloseAccount.class, this::closeAccount);
+  }
+
+  private CommandHandlerBuilder<AccountCommand, AccountEvent, ClosedAccount, Account> closedHandler() {
+    return commandHandlerBuilder(ClosedAccount.class)
+        .matchCommand(AccountCommand.class, (__, ___) -> Effect().unhandled());
+  }
+
+  @Override
+  public CommandHandler<AccountCommand, AccountEvent, Account> commandHandler() {
+    return initialHandler()
+        .orElse(openedAccountHandler())
+        .orElse(closedHandler())
+        .build();
+  }
+
+  private Effect<AccountEvent, Account> create(EmptyAccount account, CreateAccount command) {
+    return Effect().persist(new AccountCreated());
+  }
+
+  private Effect<AccountEvent, Account> deposit(OpenedAccount account, Deposit command) {
+    return Effect().persist(new Deposited(command.amount));
+  }
+
+  private Effect<AccountEvent, Account> withdraw(OpenedAccount account, Withdraw command) {
+    if (!account.canWithdraw(command.amount)) {
+      return Effect().unhandled(); // TODO replies are missing in this example
+    } else {
+      return Effect().persist(new Withdrawn(command.amount))
+        .andThen(acc2 -> { // FIXME in scaladsl it's named thenRun, change javadsl also?
+          // we know this cast is safe, but somewhat ugly
+          OpenedAccount openAccount = (OpenedAccount) acc2;
+          // do some side-effect using balance
+          System.out.println(openAccount.balance);
+        });
+    }
+  }
+
+  private Effect<AccountEvent, Account> closeAccount(OpenedAccount account, CloseAccount command) {
+    if (account.isCloseable())
+      return Effect().persist(new AccountClosed());
+    else
+      return Effect().unhandled();
+  }
+
+  @Override
+  public EventHandler<Account, AccountEvent> eventHandler() {
+    return (state, event) -> event.applyTo(state); // double-dispatch via event
+  }
```

Comment:
@TimMoore  It can be compared with doing the dispatch here instead of in the events:
```
@Override
  public EventHandler<Account, AccountEvent> eventHandler() {
    return eventHandlerBuilder()
        .matchEvent(AccountCreated.class, EmptyAccount.class, EmptyAccount::created)
        .matchEvent(Deposited.class, OpenedAccount.class, OpenedAccount::deposited)
        .matchEvent(Withdrawn.class, OpenedAccount.class, OpenedAccount::withdrawn)
        .matchEvent(AccountClosed.class, OpenedAccount.class, OpenedAccount::closed)
        .build();
  }
```

Reply:
and that doesn't require the methods to be defined in `Account`, but can be defined in the concrete state classes where they are supported

Reply:
Yeah, that looks pretty good! Maybe better to stick with that approach for now.

Reply:
With an approach like this, I also think it makes less sense for `PersistentBehavior` to be a class that you extend in Java. All of the overridden methods would return constants, right? A `PersistentBehaviorBuilder` might be more concise, and all of the interesting stuff happens in the user's state class.

Reply:
I think the purpose was mostly to capture the types once and thereby be able to provide the starting point for the builders with `this.eventHandlerBuilder()` and `this.commandHandlerBuilder()`. The builders can also be created standalone with `CommandHandlerBuilder.builder` but that might require some explicit type annotations.

Reply:
I see, yes